web_pwa_oca/controllers/main.py

```diff
@@ -1,75 +1,156 @@
-                urls.append(asset[1]['src'])
+            if asset[0] == "link":
+                urls.append(asset[1]["href"])
+            if asset[0] == "script":
+                urls.append(asset[1]["src"])
         return urls
 
-    @route('/service-worker.js', type='http', auth="public")
-    def service_worker(self):
-        qweb = request.env['ir.qweb'].sudo()
+    def _get_base_urls(self):
+        from odoo.addons.web.controllers.main import module_boot
+        url_qweb_modules = "/web/webclient/qweb?mods={}".format(",".join(module_boot()))
+        return [
+            # Cache main page
+            '/web',
+            # Cache favicon
+            "/web/static/src/img/favicon.ico",
+            # Cache manifest
+            "/web_pwa_oca/manifest.webmanifest",
+            # Cache qweb mods
+            url_qweb_modules,
+        ]
+
+    def _get_pwa_params(self):
+        """Get javascript PWA class initialzation params"""
         urls = []
-        urls.extend(self.get_asset_urls("web.assets_common"))
-        urls.extend(self.get_asset_urls("web.assets_backend"))
+        urls.extend(self._get_asset_urls("web.assets_common"))
+        urls.extend(self._get_asset_urls("web.assets_backend"))
         version_list = []
         for url in urls:
-            version_list.append(url.split('/')[3])
-        cache_version = '-'.join(version_list)
-        mimetype = 'text/javascript;charset=utf-8'
-        content = qweb.render('web_pwa_oca.service_worker', {
-            'pwa_cache_name': cache_version,
-            'pwa_files_to_cache': urls,
-        })
-        return request.make_response(content, [('Content-Type', mimetype)])
+            version_list.append(url.split("/")[3])
+        cache_version = "-".join(version_list)
+        urls.extend(self._get_base_urls())
+        return [cache_version, urls]
+
+    @route("/service-worker.js", type="http", auth="public")
+    def render_service_worker(self):
+        """Route to register the service worker in the 'main' scope ('/')"""
+        return request.render(
+            "web_pwa_oca.service_worker",
+            {
+                "pwa_scripts": self._get_pwa_scripts(),
+                "pwa_params": self._get_pwa_params(),
+            },
+            headers=[("Content-Type", "text/javascript;charset=utf-8")],
+        )
+
+    def _get_pwa_manifest_icons(self, pwa_icon):
+        icons = []
+        if not pwa_icon:
+            for size in [
+                (128, 128),
+                (144, 144),
+                (152, 152),
+                (192, 192),
+                (256, 256),
+                (512, 512),
+            ]:
+                icons.append(
+                    {
+                        "src": "/web_pwa_oca/static/img/icons/icon-%sx%s.png"
+                        % (str(size[0]), str(size[1])),
+                        "sizes": "128x128 144x144 152x152 192x192 256x256 512x512",
+                        "type": "image/svg+xml",
```

Comment:
I find it weird that the `src` ends on `png` but type is svg. Is this some required hack? If so, please add a comment.